odl/solvers/functional/default_functionals.py

```diff
@@ -570,6 +570,139 @@ def __init__(self, space):
 
 
+class QuadraticForm(Functional):
+
+    """The functional for a general quadratic form ``x^T A x + b^T x + c``."""
+
+    def __init__(self, operator=None, offset=None, constant=0):
+        """Initialize a new instance.
+
+        All parameters are optional, but at least one of ``op`` and ``offset``
+        have to be provided in turn to infer the space.
+
+        The computed value is::
+
+            x.inner(operator(x)) + offset.inner(x) + constant
+
+        Parameters
+        ----------
+        operator : `Operator`, optional
+            Linear Operator for the quadratic part of the functional.
+            ``None`` means this is ignored.
+        offset : `Operator`, optional
+            Vector for the linear part of the functional.
+            ``None`` means this is ignored.
+        constant : `Operator`, optional
+            Constant offset of the functional.
+        """
+        if operator is None and offset is None:
+            raise ValueError('need to provide at least one of ``op`` and '
+                             '``offset``')
+        if operator is not None:
+            domain = operator.domain
+        elif offset is not None:
+            domain = offset.space
+
+        if (operator is not None and offset is not None and
+                offset not in operator.domain):
+            raise ValueError('domain of `operator` and space of `offset` need '
+                             'to match')
+
+        self.__operator = operator
+        self.__offset = offset
+        self.__constant = constant
+
+        super().__init__(space=domain,
+                         linear=(operator is None and constant == 0))
+
+        if self.constant not in self.range:
+            raise ValueError('`constant` must be an element in the range of '
+                             'the functional')
+
+    @property
+    def operator(self):
+        """Operator for the quadratic part of the functional."""
+        return self.__operator
+
+    @property
+    def offset(self):
+        """Vector for the linear part of the functional."""
+        return self.__offset
+
+    @property
+    def constant(self):
+        """Constant offset of the functional."""
+        return self.__constant
+
+    def _call(self, x):
+        """Return ``self(x)``."""
+        if self.operator is None:
+            return self.offset.inner(x) + self.constant
+        elif self.offset is None:
+            return x.inner(self.operator(x)) + self.constant
+        else:
+            tmp = self.operator(x)
+            tmp += self.offset
+            return x.inner(tmp) + self.constant
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional."""
+        if self.operator is None:
+            return ConstantOperator(self.domain, self.offset)
+        else:
+            if not self.operator.is_linear:
+                # TODO: Acutally works otherwise, but needs more work
```

Comment:
It is for `F(x) = <A(x), x> + <b, x> + c`
```
d/dt_t=0 F(x + ty) = <A'(x) y, x> + <A(x), y> + <b, y> = <A'(x)^* x + A(x) + b, y>
```
i.e. the gradient is `A'(x)^* x + A(x) + b`.